content/blog/apl/2021-02-28_look_at_hpkg.md

```diff
@@ -0,0 +1,119 @@
+
+The HPKG file format is designed to facilitate this mode of use.
+
+## HPKG File Structure
+
+(_Many of the fine details of the HPKG file format are glossed over here in order to provide the reader with an overview._)
+
+### The Outer Container
+
+The structure of the "outer container" of the HPKG file is shown in the following diagram.
+
+![Outer Container](/files/blog/apl/look_at_hpkg/outercontainer.png)
+
+The "Magic" section is just a short four bytes that identify this file as being an HPKG.  This acts as a sanity check to make sure this data is most likely an HPKG file and not some other sort of a file such as an image.  A few special characters placed at the start of a file is a commonly used technique to help with detecting the type of a file.
+
+The "Header" then contains a number of critical pieces of information about the file.  The information constitutes a set of numbers that represent quantities of elements in the file, offsets into data etc... all of which is required to be able to make sense of the rest of the file.
+
+The remainder of the file, shown as a grey box in the diagram above, is called "The Heap".  At an abstract level, the heap can be considered to be a large piece of uncompressed data.  When stored, the uncompressed data is broken into 64KB pieces.  Each 64KB piece is compressed and stored as a compressed Chunk in the HPKG one after the other.  At the end of the Chunks is an array of integers that contain the compressed lengths of the Chunks.
+
+Data in the Heap is later addressed with a simple uncompressed-data offset and length.  By knowing the uncompressed chunks are (excepting the last one) 64KB in size and that the lengths of the compressed Chunks are known, it is possible to randomly access data within the Heap.
+
+### The Uncompressed Heap Structure
+
+Now that it is known roughly how the outer container works, it is possible to look into the uncompressed Heap itself without needing to consider the compression system further.
+
+![Inside Heap](/files/blog/apl/look_at_hpkg/insideheap.png)
+
+The Header provides coordinates specifying the location of two areas of the Heap used for two data structures called the Table of Contents (TOC) and the Attributes.  The TOC is a list of what file-system objects are provided for by the package.  The Attributes carry meta-data about the whole package such as the summary, description, copyrights etc...
+
+At the start of the TOC and also the Attributes resides a String Table.  Again, the Header defines the size of the String Table in each case.  The String Table is a list of strings that can be referenced from the TOC / Attributes by their index in the table.  By means of example; should the string "How Now Brown Cow" appear three times in the Attributes then it would have a single entry in the Attributes' String Table with three references to it; meaning that the string only need be stored once and not three times.
```

Comment:
```suggestion
At the start of the TOC and also the Attributes resides a String Table.  Again, the Header defines the size of the String Table in each case.  The String Table is a list of strings that can be referenced from the TOC / Attributes by their index in the table.  By means of example; should the string "How Now Brown Cow" appear three times in the Attributes then it would have a single entry in the Attributes' String Table with three references to it; meaning that the string only needs to be stored once and not three times.
```
Not 100% sure, might be valid colloquialism... 😃